beacon_node/beacon_chain/src/fetch_blobs.rs

```diff
@@ -0,0 +1,333 @@
+    RuntimeShutdown,
+}
+
+/// Fetches blobs from the EL mempool and processes them. It also broadcasts unseen blobs or
+/// data columns (PeerDAS onwards) to the network, using the supplied `publish_fn`.
+pub async fn fetch_and_process_engine_blobs<T: BeaconChainTypes>(
+    chain: Arc<BeaconChain<T>>,
+    block_root: Hash256,
+    block: Arc<SignedBeaconBlock<T::EthSpec, FullPayload<T::EthSpec>>>,
+    publish_fn: impl Fn(BlobsOrDataColumns<T::EthSpec>) + Send + 'static,
+) -> Result<Option<AvailabilityProcessingStatus>, FetchEngineBlobError> {
+    let block_root_str = format!("{:?}", block_root);
+    let log = chain
+        .log
+        .new(o!("service" => "fetch_engine_blobs", "block_root" => block_root_str));
+
+    let versioned_hashes =
+        if let Ok(kzg_commitments) = block.message().body().blob_kzg_commitments() {
+            kzg_commitments
+                .iter()
+                .map(kzg_commitment_to_versioned_hash)
+                .collect::<Vec<_>>()
+        } else {
+            debug!(
+                log,
+                "Fetch blobs not triggered - none required";
+            );
+            return Ok(None);
+        };
+
+    let num_expected_blobs = versioned_hashes.len();
+
+    let execution_layer = chain
+        .execution_layer
+        .as_ref()
+        .ok_or(FetchEngineBlobError::ExecutionLayerMissing)?;
+
+    debug!(
+        log,
+        "Fetching blobs from the EL";
+        "num_expected_blobs" => num_expected_blobs,
+    );
+    let response = execution_layer
+        .get_blobs(versioned_hashes)
+        .await
+        .map_err(FetchEngineBlobError::RequestFailed)?;
+    let num_fetched_blobs = response.iter().filter(|b| b.is_some()).count();
+
+    inc_counter_by(
+        &metrics::BLOBS_FROM_EL_EXPECTED_TOTAL,
+        num_expected_blobs as u64,
+    );
+    inc_counter_by(
+        &metrics::BLOBS_FROM_EL_RECEIVED_TOTAL,
+        num_fetched_blobs as u64,
+    );
+
+    if num_fetched_blobs == 0 {
+        debug!(
+            chain.log,
+            "No blobs fetched from the EL";
+            "num_expected_blobs" => num_expected_blobs,
+        );
+        inc_counter(&metrics::BLOBS_FROM_EL_MISS_TOTAL);
+        return Ok(None);
+    } else {
+        inc_counter(&metrics::BLOBS_FROM_EL_HIT_TOTAL);
+    }
+
+    let (signed_block_header, kzg_commitments_proof) = block
+        .signed_block_header_and_kzg_commitments_proof()
+        .map_err(FetchEngineBlobError::BeaconStateError)?;
+
+    let fixed_blob_sidecar_list = build_blob_sidecars(
+        &block,
+        response,
+        signed_block_header,
+        &kzg_commitments_proof,
+    )?;
+
+    let peer_das_enabled = chain.spec.is_peer_das_enabled_for_epoch(block.epoch());
+
+    let data_columns_receiver_opt = if peer_das_enabled {
+        // Partial blobs response isn't useful for PeerDAS, so we don't bother building and publishing data columns.
+        if !num_fetched_blobs == num_expected_blobs {
```

Comment:
Again do we work on the assumption that EL could be buggy and return more blobs than expected? I feel it's more correct to do
```suggestion
        if num_fetched_blobs < num_expected_blobs {
```

Reply:
what about making `let num_fetch_blobs = fixed_blob_sidecar_list.filter(|b| b.is_some()).count()` so we always checked against the verified and de-deuplicated list?

Reply:
This condition is using unary negation! i.e. it is calculating `!num_fetched_blobs` and then comparing that for equality with `num_expected_blobs`. I think what was intended was `!(num_fetched_blobs == num_expected_blobs)` i.e. `num_fetched_blobs != num_expected_blobs`. With the unary negation I don't think the body of this `if` would ever have been executed, because negating all the bits of a low number creates some insanely high value that is mostly 1s.

Reply:
Fixed in https://github.com/sigp/lighthouse/pull/6268/commits/756230e9589b6f15ef89495d7070c25adf3f97d3